frontend/catalyst/ag_primitives.py

```diff
@@ -411,6 +419,73 @@ def while_stmt(loop_test, loop_body, get_state, set_state, nonlocals, symbol_nam
     set_state(results)
 
 
+def _logical_op(*args, jax_fn, python_fn):
+    fallback = False
+
+    values = [f() for f in args]
+
+    def _non_scalar_tracer(x) -> bool:
+        return isinstance(x, DynamicJaxprTracer) and reduce(lambda a, b: a * b, x.shape, 1) != 1
+
+    if any(map(_non_scalar_tracer, values)):
+        raise AutoGraphError(
+            "Using logical operations (and/or/not) with non-scalar tracers is not supported\n"
+            "Consider using `jax.numpy.logical_and/..logical_or/..logical_not` functions.\n"
+            f"arguments provided were: {args}"
+        )
+
+    try:
+        if _emulate_fallback_errors:
+            raise AutoGraphError("Emulated autograph fallback error")
+
+        result = jax_fn(*values)
+    except Exception as e:  # pylint: disable=broad-exception-caught
+        if catalyst.autograph_strict_conversion:
+            raise e
+
+        import inspect
+        import textwrap
+
+        expr_info = get_source_code_info(inspect.stack()[2])
+
+        fallback = True
+
+        if not catalyst.autograph_ignore_fallbacks:
+            warnings.warn(
+                f"Tracing of an AutoGraph converted logical expression failed with an exception:\n"
+                f"  {type(e).__name__}:{textwrap.indent(str(e), '    ')}\n"
+                f"\n"
+                f"The error ocurred within the body of the following expression:\n"
+                f"{expr_info}"
+                f"\n"
+                f"To understand different types of JAX tracing errors, please refer to the "
+                f"guide at: https://jax.readthedocs.io/en/latest/errors.html\n"
+                f"\n"
+                f"If you did not intend for the conversion to happen, you may safely ignore "
+                f"this warning."
+            )
```

Comment:
I'm not sure we need any warnings here. Warnings should only exist if there is something actionable you can do about them. 

With the for loop, the warning is supposed to help users change their lists into arrays for example. 
Can you think of any cases here where the user can change their code such that it would be captured by autograph?

In fact, I think the setup of this function should be slightly different than that of the for loop. With the for loop, we _want_ to convert loops since they will be more efficient afterwards. Here, we only care that the functionality _works_.
If it already works in Python (non-array inputs), great. This is probably the preferred path. But if the inputs are abstract arrays, we _need_ the jax functions.

Reply:
Can you formulate a clear criteria when to call which path? I tried to avoid making such decisions because I don't control the implementation of `jax.numpy.logical_and/or` nor do I control `__bool__` methods of random Python objects. Simply trying to call `jax.logical_and` and only call Python's `and` on failure seems to me a more responsible decision in these circumstances.

Reply:
Another option we might follow is to _always_ use `np.logical_and/or/not`, trying to cast all arguments to bools if they are not scalar np.arrays. @dime10 what do you think?

Reply:
For reference, we decided to only convert when both operands are abstract arrays. This is the safest option because Python Boolean operators would always fail in this instance.

No fallback should be needed then.

Reply:
Implemented